### ISSUE TITLE
Consistent handling of "IPC programming errors"

### ISSUE THREAD

**Issue by cbiffle** [oxidecomputer/hubris#332]
I find it useful to distinguish between _runtime errors_ that can occur in a
correct program, and _programmer errors_ that, well, can't. There are some
interesting cases of this at the IPC boundary that don't occur in function
calls.

They all indicate that something is _very wrong_ -- most likely, a client is
misconfigured and is talking to the wrong server.

**Bad operation code.** IPC operations are distinguished by a 16-bit operation
code sent with the message. These are defined by the server's interface. If a
client sends an _unrecognized_ operation code, something is very wrong.

**Message truncation.** A server's IPC interface defines the longest message it
expects to receive. If a client sends _more_ than this, something is very wrong.

**Message that fails to decode.** The IPC interface will define some encoding
for messages -- for instance, a `bool` might be sent as 0 or 1. If a client
sends 2, something is very wrong.

**Insufficient reply space.** The IPC interface _also_ defines how long the
reply will be. If the client hasn't allocated enough space, something is very
wrong.

We signal all these cases, today, by returning an error code to the client.

It doesn't make sense to expect a client to handle these cases, because the only
reasonable thing for the client to do is panic. Putting these cases into the
error type alongside runtime errors means that the client always has to match on
them if it wants to handle any errors:

```rust
match err {
    ErrorType::Unplugged => wait_for_plug(),
    ErrorType::PowerLost => turn_power_back_on(),

    ErrorType::YourCodeIsTotallyWrong => uhhh_reconsider_life_choices(),
}
```

Down the road, these are the cases I want to handle with a `REPLY_FAULT`
operation, but I don't get the sense that anyone else is convinced this is a
good idea. So, in the interim, I'd like to propose a pattern.

Response codes from SEND are 32 bits wide. By convention, 0 is success. By
implementation, values in the range `0xFFFF_FF00..=0xFFFF_FFFF` indicate task
death.

We have historically used values 1 and 2 to indicate bad operation and bad
message size or encoding, and clients have had to deal with them.

I propose that we repurpose the values just below the dead code range to
indicate these common cases. For instance,

- `0xFFFF_FE00` as "operation not understood"
- `0xFFFF_FD00` as "bad incoming message"
- `0xFFFF_FC00` as "bad reply buffer"
- etc

And, that we _do not_ include these values in the IPC enums.

This has the implication that, on the client side, when the client goes to
decode the raw `u32` into the error enum, it will notice that something is
horribly wrong and panic. (This behavior is true today.)

(These values are numerically large to make them less likely to collide with user error enums, which start at 1 and typically count upwards from there. We could ensure that the types don't collide at the type system level by requiring user enums to provide a `Into<u16>` conversion.)

On the server side, if we need to manipulate these errors, we could "extend" any interface error type `E` with an enum like:

```rust
enum ServerError<E> {
    OperationNotUnderstood,
    BadIncomingMessage,
    BadReplyBuffer,
    Runtime(E),
}
```

...which would have an appropriate conversion to `u32`. The generated server
stubs are responsible for screening all of these cases, so server
implementations wouldn't have to see it.

In practice, these error codes will mostly be generated inside server stub support routines, so servers can probably just continue dealing in their error type `E`.

**Comment by cbiffle** [comment]
While not mentioned among the example above, two other common cases should probably get the same treatment:

- **Wrong number of leases.** Calls like SPI `write` expect exactly one lease; doing anything else means something's wrong.
- **Incorrect lease attributes.** To continue using SPI `write` as an example, its lease must be readable (because the server needs to be able to slurp bytes out of it, to send on the interface). If you call it with a lease that is not readable, something's wrong.

**Comment by cbiffle** [comment]
I feel like it's also worth noting: this proposal would not have immediate kernel implications. It only affects the IPC stubs used in tasks (client and server). _However,_ if this pattern turns out to be useful, it could _eventually_ have kernel implications. For instance, we might decide that we never want to deliver truncated messages, and have the kernel stop doing so -- instead, returning some common error code.

**Comment by cbiffle** [comment]
The changes described above have been implemented in the Idol generated servers, which covers _most_ but not all of our IPC use cases. It's been working well over the past month or so.

However,

> Down the road, these are the cases I want to handle with a `REPLY_FAULT`
> operation, but I don't get the sense that anyone else is convinced this is a
> good idea.

I have, at long last, gotten some feedback on `REPLY_FAULT` at long last, and at least one person does not think I'm being dumb. So, I have filed #368 to track. Given that the work described above is now merged, I'll close this issue and continue the conversation over there.
